### PR TITLE
Add envtest controller tests for DevWorkspaces

### PR DESCRIPTION
**Pull request by amisevsk**
### What does this PR do?
Sets up tooling and configuration to allow running controller tests via envtest. These tests work by launching a control plane and allowing the controller to "interact" with it -- i.e. creating DevWorkspaces and related objects.

Currently a draft PR as this is a revival of something I started back in January and hadn't touched again until a few weeks ago; I'm not sure the overall quality.

Tests can still be run without installing the ginkgo CLI, but the output is nicer if it's present.

Controller-gen assumes a newer version of ginkgo/gomega than was used previously; these packages are updated and that requires updates to existing e2e test code.

Docs: 
* Kubebuilder controller tests: https://book.kubebuilder.io/cronjob-tutorial/writing-tests.html
* Envtest setup: https://book.kubebuilder.io/reference/envtest.html
* Ginkgo docs: https://onsi.github.io/ginkgo/
* Gomega docs: https://onsi.github.io/gomega/

### What issues does this PR fix or reference?
Closes https://github.com/devfile/devworkspace-operator/issues/930

### Is it tested? How?
They're tests :) 

### PR Checklist

- [ ] E2E tests pass (when PR is ready, comment `/test v8-devworkspace-operator-e2e, v8-che-happy-path` to trigger)
    - [ ] `v8-devworkspace-operator-e2e`: DevWorkspace e2e test
    - [ ] `v8-che-happy-path`: Happy path for verification integration with Che

### CODE REVIEW

**Review by AObuchow — COMMENTED** [review]
Looks awesome to me so far :) 

One additional note:
The README should probably be updated to specify that go v1.17 is required, as well as Ginkgo v2.0.0.

File: controllers/workspace/util_test.go
Comment (AObuchow) [reply controllers/workspace/util_test.go:318]: Let me know if you think this is worthwhile, and I'll create an issue about it.

Comment (amisevsk) [reply controllers/workspace/util_test.go:318]: Eh, I think it's probably not worth a separate package that has to be imported everywhere. Golang is a language that includes code duplication by design :shrug:

File: .github/workflows/pr.yml
Comment (amisevsk) [reply .github/workflows/pr.yml:115]: Note that the Codecov action was also disabled for PRs; previously we were running the tests as part of the PR check and generating coverage files, and then running them again as part of the Codecov action; it's easier to just upload the already-generated files.

I should break this out into a separate PR perhaps; maybe that would fix the codecov weirdness?

File: controllers/workspace/util_test.go
Comment (amisevsk) [reply controllers/workspace/util_test.go]: Good idea -- updated the PR.

File: controllers/workspace/util_test.go
Comment (AObuchow) [reply controllers/workspace/util_test.go:317]: > probably not worth a separate package that has to be imported everywhere

Agreed +1 

> Golang is a language that includes code duplication by design :shrug:

Very true ;P

File: .github/workflows/pr.yml
Comment (AObuchow) [reply .github/workflows/pr.yml:115]: > it's easier to just upload the already-generated files

Agreed, it'll speed things up. If we do this, it's probably worth removing the [PR trigger](https://github.com/devfile/devworkspace-operator/blob/main/.github/workflows/code-coverage.yml#L7) for the code coverage action (and only have it run on push to main), cause otherwise it will get run twice for PR's (I would imagine?).

> I should break this out into a separate PR perhaps; maybe that would fix the codecov weirdness?

It should fix the weirdness because then there will be a reference point of coverage % for `controller.cover.out` on the main branch. But if this causes other problems, then it's totally fine to leave it. Future PR's should not have this issue once the `controller.cover.out` is uploaded to codecov.

Comment (amisevsk) [reply .github/workflows/pr.yml:115]: > Agreed, it'll speed things up. If we do this, it's probably worth removing the [PR trigger](https://github.com/devfile/devworkspace-operator/blob/main/.github/workflows/code-coverage.yml#L7) for the code coverage action (and only have it run on push to main), cause otherwise it will get run twice for PR's (I would imagine?).

I thought I had already done this -- https://github.com/devfile/devworkspace-operator/pull/957/files#diff-49708f979e226a1e7bd7a68d71b2e91aae8114dd3e9254d9830cd3b4d62d4303L7-L8. Is anything else required?

My understanding is that codecov merges all uploaded files: https://docs.codecov.com/docs/merging-reports. I think this means that it should correctly compute a diff even if we were to e.g. split coverage reports per package.

Comment (AObuchow) [reply .github/workflows/pr.yml:115]: > I thought I had already done this -- [#957 (files)](https://github.com/devfile/devworkspace-operator/pull/957/files#diff-49708f979e226a1e7bd7a68d71b2e91aae8114dd3e9254d9830cd3b4d62d4303L7-L8). Is anything else required?

Oops I forgot about that when I wrote this - that should be all that's needed for the code coverage action.
 
> My understanding is that codecov merges all uploaded files: [docs.codecov.com/docs/merging-reports](https://docs.codecov.com/docs/merging-reports). I think this means that it should correctly compute a diff even if we were to e.g. split coverage reports per package.

Yes, that seems to be the case. My theory about the `controller.cover.out` not being present on master doesn't seem to hold up then. 

I think it's okay to just include the CI changes in this PR and if codecov continues to display incorrect coverage reports after this PR is merged, I will look into it.

**Review by AObuchow — APPROVED** [review]

**Review by ibuziuk — APPROVED** [review]

**Pull request merged**
